File: web/src/api/k8s/workloads/daemonSet.ts
import { request } from "@/utils/service"
import { commonListReq, type commonRelatedReq, type commonReq } from "../entry"

// 获取daemonSet list
export interface DaemonSetBrief {
  name: string
  namespace: string
  pods: string
  creationTimestamp: string
}

interface DaemonSetBriefList extends PageInfo {
  list: DaemonSetBrief[]
  total: number
}

export const getDaemonSetsApi = (data: commonListReq) => {
  return request<ApiResponseData<DaemonSetBriefList>>({
    url: "/k8s/daemonSet/getDaemonSets",
    method: "post",
    data
  })
}

// 获取daemonSet详情

export interface DaemonSetDetail {
  metadata: {}
  spec: {
    selector: object
    updateStrategy: { type: string; rollingUpdate: { maxUnavailable: number; maxSurge: number } }
  }
  status: {
    currentNumberScheduled: number
    numberMisscheduled: number
    desiredNumberScheduled: number
    numberReady: number
    updatedNumberScheduled: number
    numberAvailable: number
    conditions: []
  }
}

export const getDaemonSetDetailApi = (data: commonReq) => {
  return request<ApiResponseData<DaemonSetDetail>>({
    url: "/k8s/daemonSet/getDaemonSetDetail",
    method: "post",
    data
  })
}

// 删除daemonSet
export const deleteDaemonSetApi = (data: commonReq) => {
  return request<ApiResponseData<null>>({
    url: "/k8s/daemonSet/deleteDaemonSet",
    method: "post",
    data
  })
}

// 获取daemonSet关联的pod

export const getDaemonSetPodsApi = (data: commonRelatedReq) => {
  return request<ApiResponseData<any>>({
    url: "/k8s/daemonSet/getDaemonSetPods",
    method: "post",
    data
  })
}

File: server/model/k8s/workloads/daemonSet.go
package workloads

import (
	"github.com/pddzl/kubefish/server/model/k8s"
	"k8s.io/apimachinery/pkg/util/intstr"
	"time"
)

// DaemonSetBrief for list
type DaemonSetBrief struct {
	Name              string    `json:"name"`
	NameSpace         string    `json:"namespace"`
	Pods              string    `json:"pods"`
	CreationTimestamp time.Time `json:"creationTimestamp,omitempty"`
}

// DaemonSetDetail for detail
type DaemonSetDetail struct {
	ObjectMeta k8s.ObjectMeta  `json:"metadata,omitempty"`
	Spec       daemonSetSpec   `json:"spec,omitempty"`
	Status     daemonSetStatus `json:"status,omitempty"`
}

type daemonSetSpec struct {
	Selector             map[string]string       `json:"selector"`
	UpdateStrategy       daemonSetUpdateStrategy `json:"updateStrategy,omitempty"`
	MinReadySeconds      int32                   `json:"minReadySeconds,omitempty"`
	RevisionHistoryLimit *int32                  `json:"revisionHistoryLimit,omitempty"`
}

type daemonSetStatus struct {
	CurrentNumberScheduled int32                `json:"currentNumberScheduled"`
	NumberMisscheduled     int32                `json:"numberMisscheduled"`
	DesiredNumberScheduled int32                `json:"desiredNumberScheduled"`
	NumberReady            int32                `json:"numberReady"`
	UpdatedNumberScheduled int32                `json:"updatedNumberScheduled,omitempty"`
	NumberAvailable        int32                `json:"numberAvailable,omitempty"`
	NumberUnavailable      int32                `json:"numberUnavailable,omitempty"`
	Conditions             []DaemonSetCondition `json:"conditions,omitempty"`
}

type DaemonSetCondition struct {
	Type               string    `json:"type"`
	Status             string    `json:"status"`
	LastTransitionTime time.Time `json:"lastTransitionTime,omitempty"`
	Reason             string    `json:"reason,omitempty"`
	Message            string    `json:"message,omitempty"`
}

type daemonSetUpdateStrategy struct {
	Type          string                 `json:"type,omitempty"`
	RollingUpdate rollingUpdateDaemonSet `json:"rollingUpdate,omitempty"`
}

type rollingUpdateDaemonSet struct {
	MaxUnavailable *intstr.IntOrString `json:"maxUnavailable,omitempty"`
	MaxSurge       *intstr.IntOrString `json:"maxSurge,omitempty"`
}
